### PR TITLE
[query]  NDArray Based Logistic Regression Round 1

### PR DESCRIPTION
**Pull request by johnc1231**
This PR reimplements the Wald and LRT test types of hail's logistic regression rows in terms of ndarrays. It is not yet in a state where it should replace hail's current Breeze based logistic regression since:

1. I still have to implement the `'firth'` and `'score'` test types (there's some extra work to implement Firth specifically that's going to mean adding a new ndarray method)
2. It is slower than Hail's Breeze based logistic regression.

I'd still like to get it in its current state, as it's still a feature add to the local and service backends, and I intend to follow up by adding a benchmark for this to track performance changes. 

This PR also fixes two bugs I found along the way:

1. In Python, `TailLoop._compute_type` was not recursively calling `_compute_type` on its params, which was a mistake. 
2. TailLoop creates bindings, but didn't have a `PruneDeadFields` rule. I added one based on the `Let` rule we already had.

### CODE REVIEW

**Review by patrick-schultz — DISMISSED** [review]
Nice work John!

File: hail/python/hail/methods/statgen.py [hunk 1]
```diff
@@ -815,6 +815,415 @@ def logistic_regression_rows(test, y, x, covariates, pass_through=()) -> hail.Ta
...
+        delta_b = delta_b_struct.solution
+        max_delta_b = nd_max(delta_b.map(lambda e: hl.abs(e)))
+        log_lkhd = ((y * mu) + (1 - y) * (1 - mu)).map(lambda e: hl.log(e)).sum()
+
+        def compute_next_iter(cur_iter, b, mu, score, fisher):
+            cur_iter = cur_iter + 1
+            b = b + delta_b
+            mu = sigmoid(X @ b)
+            score = X.T @ (y - mu)
+            fisher = X.T @ (X * (mu * (1 - mu)).reshape(-1, 1))
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/python/hail/methods/statgen.py:890]: Thanks, will look into this / maybe discuss at linalg meeting.

File: hail/python/hail/methods/statgen.py [hunk 2]
```diff
@@ -815,6 +815,415 @@ def logistic_regression_rows(test, y, x, covariates, pass_through=()) -> hail.Ta
...
+def logreg_fit(X, y, null_fit=None, max_iter=25, tol=1E-6):
+    assert(X.ndim == 2)
+    assert(y.ndim == 1)
+    # X is samples by covs.
+    # y is length num samples, for one cov.
+    n = X.shape[0]
+    m = X.shape[1]
+
+    if null_fit is None:
+        b = hl.nd.array(hl.array([y.sum()/n]).extend(hl.zeros(hl.int32(m - 1)).map(lambda e: hl.float64(e))))
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/python/hail/methods/statgen.py]: You're right, fixed.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

File: hail/python/hail/methods/statgen.py [hunk 3]
```diff
@@ -815,6 +815,415 @@ def logistic_regression_rows(test, y, x, covariates, pass_through=()) -> hail.Ta
...
+def logreg_fit(X, y, null_fit=None, max_iter=25, tol=1E-6):
+    assert(X.ndim == 2)
+    assert(y.ndim == 1)
+    # X is samples by covs.
+    # y is length num samples, for one cov.
+    n = X.shape[0]
+    m = X.shape[1]
+
+    if null_fit is None:
+        b = hl.nd.hstack([hl.nd.array([hl.log(y.sum() / n)]), hl.nd.zeros((hl.int32(m - 1)))])
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/python/hail/methods/statgen.py]: Fixed, thanks for checking thoroughly. I hate that tests don't catch this.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**